### PR TITLE
feat: create rust SDK

### PR DESCRIPTION
**Pull request by 0xE282B0**
First commit for the rust-sdk:

- Initial bindgen
- first example using unsafe code

based on discussion in #2684

### CODE REVIEW

File: language-bindings/rust/build.rs [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+
+use cmake::Config;
+use std::{env, path::PathBuf};
+
+fn main() {
+    let mut config = Config::new("../..");
+    let dst = config
+        .generator("Unix Makefiles")
+        .define("WAMR_BUILD_TARGET", env::var("TARGET").unwrap().split("-").next().unwrap().to_uppercase())
+        .build();
```
Comment (lum1n0us) [inline language-bindings/rust/build.rs:12]: please feel free to correct me.

becuase we can't ask crate users to re-compile,
I guess we need to prepare a compilation options combination here in order to define 
- which features we are going to provide in the final publish crate 
- which running mode we are going to provide in the final publish crate.

Comment (0xE282B0) [reply language-bindings/rust/build.rs:12]: The build.rs is run during the crate installation. It is not uncommon to compile dependencies here.

For the execution mode (like interpreted, AOT, JIT) we can hopefully use feature flags on the crate.

For environments where it is not possible to compile wamr, we could try to "vendor" wamr and include it in the crate.

File: language-bindings/rust/build.rs [hunk 2]
```diff
@@ -0,0 +1,30 @@
+extern crate bindgen;
```
Comment (wenyongh) [inline language-bindings/rust/build.rs:1]: Should add the license header, note that the license must be "Apache-2.0 WITH LLVM-exception", refer to:
https://github.com/bytecodealliance/wasm-micro-runtime/blob/main/core/iwasm/libraries/lib-wasi-threads/lib_wasi_threads_wrapper.c#L2-L3
https://github.com/bytecodealliance/wasm-micro-runtime/blob/main/core/iwasm/libraries/debug-engine/debug_engine.c#L2-L3

File: language-bindings/rust/samples/wasi-hello/src/main.rs [hunk 3]
```diff
@@ -0,0 +1,123 @@
+use wamr_sdk_rust::*;
```
Comment (wenyongh) [inline language-bindings/rust/samples/wasi-hello/src/main.rs:1]: Should add license header.

File: language-bindings/rust/build.rs (hunk 2 above)
Comment (0xE282B0) [reply language-bindings/rust/build.rs:1]: Sure, makes sense. Does the license header look correct now?
I also added the license to the `Cargo.toml`.

Comment (wenyongh) [reply language-bindings/rust/build.rs:6]: Yes, it looks good. BTW, I didn't find license header in Cargo.toml of this folder (it is added the file of another folder), could you add it?

File: language-bindings/rust/Cargo.toml [hunk 4]
```diff
@@ -0,0 +1,13 @@
+[package]
+name = "wamr-sdk-rust"
+version = "0.1.0"
+edition = "2021"
```
Comment (wenyongh) [inline language-bindings/rust/Cargo.toml:4]: Is it better to change to 2023?

File: language-bindings/rust/samples/wasi-hello/Cargo.toml [hunk 5]
```diff
@@ -0,0 +1,12 @@
+[package]
+name = "wasi-hello"
+version = "0.1.0"
+edition = "2021"
```
Comment (wenyongh) [inline language-bindings/rust/samples/wasi-hello/Cargo.toml:4]: Is it better to change to 2023?

File: language-bindings/rust/build.rs [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+extern crate cmake;
+
+use cmake::Config;
+use std::{env, path::PathBuf};
+
+fn main() {
+    let mut config = Config::new("../..");
+    let dst = config
+        .generator("Unix Makefiles")
+        .define("WAMR_BUILD_TARGET", env::var("TARGET").unwrap().split("-").next().unwrap().to_uppercase())
```
Comment (lum1n0us) [inline language-bindings/rust/build.rs:16]: I think we'd better let CMakeFiles to decide the value of `WAMR_BUILD_TAEGET`.

Plus, I am not sure `-G 'Unix Makefiles'` is suitable here. at least, as I know, it won't work on Windows(not with mingw)

File: language-bindings/rust/Cargo.toml [hunk 7]
```diff
@@ -0,0 +1,13 @@
...
+
```
Comment (lum1n0us) [inline language-bindings/rust/Cargo.toml:5]: so there will be a `[features]` section here?

File: language-bindings/rust/samples/wasi-hello/src/main.rs [hunk 8]
```diff
@@ -0,0 +1,128 @@
...
+            error_buf.len() as u32,
+        )
+    };
+
+    //let err_u8_vec: Vec<u8> = error_buf.iter().map(|&x| x as u8).filter(|x| *x > 31 && *x < 124).collect();
+    //print!("error {:?}", String::from_utf8(err_u8_vec));
+
+    assert!((module_instance as usize) != 0);
+
+    let _exec_env = unsafe { wasm_runtime_create_exec_env(module_instance, DEFAULT_STACK_SIZE) };
```
Comment (wenyongh) [inline language-bindings/rust/samples/wasi-hello/src/main.rs:120]: No need to create exec_env if you just to want to call wasm_application_execute_main, in which exec_env is created automatically.

File: language-bindings/rust/samples/wasi-hello/src/main.rs [hunk 9]
```diff
@@ -0,0 +1,128 @@
...
+
+    let success =
+        unsafe { wasm_application_execute_main(module_instance, 0, std::ptr::null_mut()) };
+
+    assert!(success);
+
+    let _main_result = unsafe { wasm_runtime_get_wasi_exit_code(module_instance) };
```
Comment (wenyongh) [inline language-bindings/rust/samples/wasi-hello/src/main.rs:127]: Not sure whether we should wasm_runtime_deinstantiate(module_inst), wasm_runtime_unload(module) and wasm_runtime_destroy() before return in rust?
See
https://github.com/bytecodealliance/wasm-micro-runtime/blob/main/product-mini/platforms/posix/main.c#L978-L999

File: language-bindings/rust/Cargo.toml (hunk 4 above)
Comment (0xE282B0) [reply language-bindings/rust/Cargo.toml:4]: Editions in rust are used to introduce backward incompatible changes without breaking existing projects; a new edition is not created every year. Currently there are three editions, `2015`, `2018`, and `2021`. New crates created by cargo new are always configured to use the latest stable edition.

There is a guide to rust editions at https://doc.rust-lang.org/edition-guide/editions/index.html

File: language-bindings/rust/samples/wasi-hello/src/main.rs (hunk 8 above)
Comment (0xE282B0) [reply language-bindings/rust/samples/wasi-hello/src/main.rs:120]: Good to know! I'm not very familiar with the API.
I just realize that the return value is never used. 😅

File: language-bindings/rust/build.rs (hunk 2 above)
Comment (0xE282B0) [reply language-bindings/rust/build.rs:6]: Oh, I meant I added the license information field:
```toml
  license = "Apache-2.0 WITH LLVM-exception"
```

But I can also leave a license header there 😊

File: language-bindings/rust/samples/wasi-hello/src/main.rs (hunk 9 above)
Comment (0xE282B0) [reply language-bindings/rust/samples/wasi-hello/src/main.rs:127]: You're right, even if this is just an example, it should clean up in the end.

File: language-bindings/rust/build.rs (hunk 6 above)
Comment (0xE282B0) [reply language-bindings/rust/build.rs:16]: RE build target: this is a leftover from the crate I forked, I had not tried to remove it.
RE generator: I just have no experience with this, I haven't used windows or mingw in years.

The build process seems to work fine without these two configs, so I remove them.

File: language-bindings/rust/build.rs (hunk 2 above)
Comment (0xE282B0) [reply language-bindings/rust/build.rs:6]: Oh, right, my bad. There are two `Cargo.toml` and now they should  both have a License header and a license field for the crate:
```toml
  license = "Apache-2.0 WITH LLVM-exception"
```

File: language-bindings/rust/Cargo.toml (hunk 4 above)
Comment (wenyongh) [reply language-bindings/rust/Cargo.toml:7]: Got it, thanks.

File: language-bindings/rust/samples/wasi-hello/src/main.rs [hunk 10]
```diff
@@ -0,0 +1,131 @@
...
+
+    unsafe {
+        wasm_runtime_deinstantiate(module_instance);
+        wasm_runtime_unload(module);
```
Comment (wenyongh) [inline language-bindings/rust/samples/wasi-hello/src/main.rs:129]: Should also add `wasm_runtime_destroy()` at the end.

**Review by wenyongh — COMMENTED** [review]
LGTM

**Comment by lum1n0us** [comment]
LGTM

**Pull request merged**